[flang] Fix homonymous interface and procedure warning

luporl opened the pull request:
When emitting an homonymous generic interface and procedure warning,
the source locations of the interface and the procedure were being
compared to find the one that occurred later in the source file.

The problem is that they could be in different source/module files,
which makes the comparison invalid.

Fix it by using parser::AllCookedSources::Precedes() instead, that
correctly handle names in different source files.

luporl committed:
[flang] Make lto-bc.f90 unsupported on Darwin

On Darwin, an additional unexpected warning is emitted:

```
/path/to/flang/test/Semantics/bug159554.f90:4:11: warning: 'c_funloc' should not be the name of both a generic interface and a procedure unless it is a specific procedure of the generic [-Whomonymous-specific]
  interface c_funloc
            ^^^^^^^^
```

eugeneepshteyn commented:
@jeanPerier , this is the weird failure that shows up on some buildbots, but not the others, right? Like the Windows buildbot?

@luporl , is it consistent on Darwin? Any chance you could do some debugging?

eugeneepshteyn commented:
Even a call stack of how it got to that warning may be helpful.

luporl committed:
[flang] Fix homonymous interface and procedure warning

When emitting an homonymous generic interface and procedure warning,
the source locations of the interface and the procedure were being
compared to find the one that occurred later in the source file.

The problem is that they could be in different source/module files,
which makes the comparison invalid.

Fix it by prioritizing the current source file over module files and
by comparing the source locations only when both the generic interface
and the procedure are not in module files.

luporl committed:
Revert "[flang] Make lto-bc.f90 unsupported on Darwin"

This reverts commit 7c7c07f57ef00528b09cb4f647ee6eb0eccdb142.

luporl commented:
The reason why the warning didn't appear in some cases was because in those cases the source location of a module file was being selected, which makes `Warn()` skip the message.

On Linux the contents of the iso_c_binding module were usually in a higher memory location than the contents of the current source file, which made the location of the module be selected, causing the warning to be skipped.
On Darwin it was the other way around.

klausler commented on flang/lib/Semantics/resolve-names.cpp:
```diff
@@ -4535,9 +4535,24 @@ void InterfaceVisitor::CheckGenericProcedures(Symbol &generic) {
-    context().Warn(common::UsageWarning::HomonymousSpecific,
-        proc->name().begin() > generic.name().begin() ? proc->name()
-                                                      : generic.name(),
+    // Use the source location that is not in a module file for the warning,
```
There's already infrastructure in the compiler that produces reproducible results for name comparisons, and it should have been used here.  See `parser::AllCookedSources::Precedes(CharBlock x, CharBlock y)`, or its client `semantics::SymbolSourcePositionCompare::operator()`.

luporl committed:
Use AllCookedSources::Precedes to compare source locations
